// @storybook/components/dist/tap-PX6LWER6.mjs
import {
  __commonJS,
  init_define_process_env
} from "./chunk-NIB2YEFO.mjs";

// ../../node_modules/highlight.js/lib/languages/tap.js
var require_tap = __commonJS({
  "../../node_modules/highlight.js/lib/languages/tap.js"(exports, module) {
    init_define_process_env();
    function tap(hljs) {
      return {
        name: "Test Anything Protocol",
        case_insensitive: true,
        contains: [
          hljs.HASH_COMMENT_MODE,
          {
            className: "meta",
            variants: [
              {
                begin: "^TAP version (\\d+)$"
              },
              {
                begin: "^1\\.\\.(\\d+)$"
              }
            ]
          },
          {
            begin: /---$/,
            end: "\\.\\.\\.$",
            subLanguage: "yaml",
            relevance: 0
          },
          {
            className: "number",
            begin: " (\\d+) "
          },
          {
            className: "symbol",
            variants: [
              {
                begin: "^ok"
              },
              {
                begin: "^not ok"
              }
            ]
          }
        ]
      };
    }
    module.exports = tap;
  }
});
export default require_tap();
