// @storybook/addon-interactions/dist/preset/checkActionsLoaded.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/preset/checkActionsLoaded.ts
var checkActionsLoaded_exports = {};
__export(checkActionsLoaded_exports, {
  checkActionsLoaded: () => checkActionsLoaded
});
module.exports = __toCommonJS(checkActionsLoaded_exports);
var import_core_common = require("@storybook/core-common");
var import_path = __toESM(require("path"));
var checkActionsLoaded = (configDir) => {
  (0, import_core_common.checkAddonOrder)({
    before: {
      name: "@storybook/addon-actions",
      inEssentials: true
    },
    after: {
      name: "@storybook/addon-interactions",
      inEssentials: false
    },
    configFile: import_path.default.isAbsolute(configDir) ? import_path.default.join(configDir, "main") : import_path.default.join(process.cwd(), configDir, "main"),
    getConfig: (configFile) => (0, import_core_common.serverRequire)(configFile)
  });
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  checkActionsLoaded
});
