// @storybook/preview-web/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DocsContext", {
  enumerable: true,
  get: function () {
    return _DocsContext.DocsContext;
  }
});
Object.defineProperty(exports, "Preview", {
  enumerable: true,
  get: function () {
    return _Preview.Preview;
  }
});
Object.defineProperty(exports, "PreviewWeb", {
  enumerable: true,
  get: function () {
    return _PreviewWeb.PreviewWeb;
  }
});
Object.defineProperty(exports, "composeConfigs", {
  enumerable: true,
  get: function () {
    return _store.composeConfigs;
  }
});
Object.defineProperty(exports, "simulateDOMContentLoaded", {
  enumerable: true,
  get: function () {
    return _simulatePageload.simulateDOMContentLoaded;
  }
});
Object.defineProperty(exports, "simulatePageLoad", {
  enumerable: true,
  get: function () {
    return _simulatePageload.simulatePageLoad;
  }
});

var _store = require("@storybook/store");

var _Preview = require("./Preview");

var _PreviewWeb = require("./PreviewWeb");

var _simulatePageload = require("./simulate-pageload");

var _DocsContext = require("./docs-context/DocsContext");
