// @storybook/addon-links/dist/react/index.mjs
import {
  hrefTo,
  navigate
} from "../chunk-MC5INVAQ.mjs";
import "../chunk-5CZDYUKT.mjs";

// src/react/components/link.tsx
import React, { PureComponent } from "react";
var LEFT_BUTTON = 0;
var isPlainLeftClick = (e) => e.button === LEFT_BUTTON && !e.altKey && !e.ctrlKey && !e.metaKey && !e.shiftKey;
var cancelled = (e, cb = (_e) => {
}) => {
  if (isPlainLeftClick(e)) {
    e.preventDefault();
    cb(e);
  }
};
var LinkTo = class extends PureComponent {
  constructor() {
    super(...arguments);
    this.state = {
      href: "/"
    };
    this.updateHref = async () => {
      const { kind, story } = this.props;
      if (kind && story) {
        const href = await hrefTo(kind, story);
        this.setState({ href });
      }
    };
    this.handleClick = () => {
      const { kind, story } = this.props;
      if (kind && story) {
        navigate({ kind, story });
      }
    };
  }
  componentDidMount() {
    this.updateHref();
  }
  componentDidUpdate(prevProps) {
    const { kind, story } = this.props;
    if (prevProps.kind !== kind || prevProps.story !== story) {
      this.updateHref();
    }
  }
  render() {
    const { kind, story, children, ...rest } = this.props;
    const { href } = this.state;
    return /* @__PURE__ */ React.createElement("a", {
      href,
      onClick: (e) => cancelled(e, this.handleClick),
      ...rest
    }, children);
  }
};
LinkTo.defaultProps = {
  kind: null,
  story: null,
  children: void 0
};

// src/react/index.ts
var react_default = LinkTo;
export {
  react_default as default
};
