// @storybook/node-logger/dist/index.mjs
// src/index.ts
import npmLog from "npmlog";
import prettyTime from "pretty-hrtime";
import chalk from "chalk";
var colors = {
  pink: chalk.hex("F1618C"),
  purple: chalk.hex("B57EE5"),
  orange: chalk.hex("F3AD38"),
  green: chalk.hex("A2E05E"),
  blue: chalk.hex("6DABF5"),
  red: chalk.hex("F16161"),
  gray: chalk.gray
};
var logger = {
  verbose: (message) => npmLog.verbose("", message),
  info: (message) => npmLog.info("", message),
  plain: (message) => console.log(message),
  line: (count = 1) => console.log(`${Array(count - 1).fill("\n")}`),
  warn: (message) => npmLog.warn("", message),
  error: (message) => npmLog.error("", message),
  trace: ({ message, time }) => npmLog.info("", `${message} (${colors.purple(prettyTime(time))})`),
  setLevel: (level = "info") => {
    npmLog.level = level;
  }
};
var logged = /* @__PURE__ */ new Set();
var once = (type) => (message) => {
  if (logged.has(message))
    return void 0;
  logged.add(message);
  return logger[type](message);
};
once.clear = () => logged.clear();
once.verbose = once("verbose");
once.info = once("info");
once.warn = once("warn");
once.error = once("error");
export {
  colors,
  npmLog as instance,
  logger,
  once
};
