// @storybook/addon-links/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  LinkTo: () => LinkTo,
  hrefTo: () => hrefTo,
  linkTo: () => linkTo,
  navigate: () => navigate,
  withLinks: () => withLinks
});
module.exports = __toCommonJS(src_exports);
var import_ts_dedent = require("ts-dedent");

// src/utils.ts
var import_global = __toESM(require("global"));
var import_addons = require("@storybook/addons");
var import_core_events = require("@storybook/core-events");
var import_csf = require("@storybook/csf");

// src/constants.ts
var ADDON_ID = "storybook/links";
var PARAM_KEY = `links`;
var constants_default = {
  NAVIGATE: `${ADDON_ID}/navigate`,
  REQUEST: `${ADDON_ID}/request`,
  RECEIVE: `${ADDON_ID}/receive`
};

// src/utils.ts
var { document, HTMLElement } = import_global.default;
function parseQuery(queryString) {
  const query = {};
  const pairs = (queryString[0] === "?" ? queryString.substring(1) : queryString).split("&").filter(Boolean);
  for (let i = 0; i < pairs.length; i++) {
    const pair = pairs[i].split("=");
    query[decodeURIComponent(pair[0])] = decodeURIComponent(pair[1] || "");
  }
  return query;
}
var navigate = (params) => import_addons.addons.getChannel().emit(import_core_events.SELECT_STORY, params);
var hrefTo = (title, name) => {
  return new Promise((resolve) => {
    const { location } = document;
    const query = parseQuery(location.search);
    const existingId = [].concat(query.id)[0];
    const titleToLink = title || existingId.split("--", 2)[0];
    const id = (0, import_csf.toId)(titleToLink, name);
    const url = `${location.origin + location.pathname}?${Object.entries({ ...query, id }).map((item) => `${item[0]}=${item[1]}`).join("&")}`;
    resolve(url);
  });
};
var valueOrCall = (args) => (value) => typeof value === "function" ? value(...args) : value;
var linkTo = (idOrTitle, nameInput) => (...args) => {
  const resolver = valueOrCall(args);
  const title = resolver(idOrTitle);
  const name = nameInput ? resolver(nameInput) : false;
  if ((title == null ? void 0 : title.match(/--/)) && !name) {
    navigate({ storyId: title });
  } else if (name && title) {
    navigate({ kind: title, story: name });
  } else if (title) {
    navigate({ kind: title });
  } else if (name) {
    navigate({ story: name });
  }
};
var linksListener = (e) => {
  const { target } = e;
  if (!(target instanceof HTMLElement)) {
    return;
  }
  const element = target;
  const { sbKind: kind, sbStory: story } = element.dataset;
  if (kind || story) {
    e.preventDefault();
    navigate({ kind, story });
  }
};
var hasListener = false;
var on = () => {
  if (!hasListener) {
    hasListener = true;
    document.addEventListener("click", linksListener);
  }
};
var off = () => {
  if (hasListener) {
    hasListener = false;
    document.removeEventListener("click", linksListener);
  }
};
var withLinks = (0, import_addons.makeDecorator)({
  name: "withLinks",
  parameterName: PARAM_KEY,
  wrapper: (getStory, context) => {
    on();
    import_addons.addons.getChannel().once(import_core_events.STORY_CHANGED, off);
    return getStory(context);
  }
});

// src/index.ts
var hasWarned = false;
function LinkTo() {
  if (!hasWarned) {
    console.error(import_ts_dedent.dedent`
      LinkTo has moved to addon-links/react:
      import LinkTo from '@storybook/addon-links/react';
    `);
    hasWarned = true;
  }
  return null;
}
if (module && module.hot && module.hot.decline) {
  module.hot.decline();
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LinkTo,
  hrefTo,
  linkTo,
  navigate,
  withLinks
});
