// @storybook/components/dist/smalltalk-2ZAJCL7R.mjs
import {
  __commonJS,
  init_define_process_env
} from "./chunk-NIB2YEFO.mjs";

// ../../node_modules/highlight.js/lib/languages/smalltalk.js
var require_smalltalk = __commonJS({
  "../../node_modules/highlight.js/lib/languages/smalltalk.js"(exports, module) {
    init_define_process_env();
    function smalltalk(hljs) {
      const VAR_IDENT_RE = "[a-z][a-zA-Z0-9_]*";
      const CHAR = {
        className: "string",
        begin: "\\$.{1}"
      };
      const SYMBOL = {
        className: "symbol",
        begin: "#" + hljs.UNDERSCORE_IDENT_RE
      };
      return {
        name: "Smalltalk",
        aliases: ["st"],
        keywords: "self super nil true false thisContext",
        contains: [
          hljs.COMMENT('"', '"'),
          hljs.APOS_STRING_MODE,
          {
            className: "type",
            begin: "\\b[A-Z][A-Za-z0-9_]*",
            relevance: 0
          },
          {
            begin: VAR_IDENT_RE + ":",
            relevance: 0
          },
          hljs.C_NUMBER_MODE,
          SYMBOL,
          CHAR,
          {
            begin: "\\|[ ]*" + VAR_IDENT_RE + "([ ]+" + VAR_IDENT_RE + ")*[ ]*\\|",
            returnBegin: true,
            end: /\|/,
            illegal: /\S/,
            contains: [{
              begin: "(\\|[ ]*)?" + VAR_IDENT_RE
            }]
          },
          {
            begin: "#\\(",
            end: "\\)",
            contains: [
              hljs.APOS_STRING_MODE,
              CHAR,
              hljs.C_NUMBER_MODE,
              SYMBOL
            ]
          }
        ]
      };
    }
    module.exports = smalltalk;
  }
});
export default require_smalltalk();
