// @storybook/components/dist/taggerscript-IZSH7HSW.mjs
import {
  __commonJS,
  init_define_process_env
} from "./chunk-NIB2YEFO.mjs";

// ../../node_modules/highlight.js/lib/languages/taggerscript.js
var require_taggerscript = __commonJS({
  "../../node_modules/highlight.js/lib/languages/taggerscript.js"(exports, module) {
    init_define_process_env();
    function taggerscript(hljs) {
      const COMMENT = {
        className: "comment",
        begin: /\$noop\(/,
        end: /\)/,
        contains: [{
          begin: /\(/,
          end: /\)/,
          contains: [
            "self",
            {
              begin: /\\./
            }
          ]
        }],
        relevance: 10
      };
      const FUNCTION = {
        className: "keyword",
        begin: /\$(?!noop)[a-zA-Z][_a-zA-Z0-9]*/,
        end: /\(/,
        excludeEnd: true
      };
      const VARIABLE = {
        className: "variable",
        begin: /%[_a-zA-Z0-9:]*/,
        end: "%"
      };
      const ESCAPE_SEQUENCE = {
        className: "symbol",
        begin: /\\./
      };
      return {
        name: "Tagger Script",
        contains: [
          COMMENT,
          FUNCTION,
          VARIABLE,
          ESCAPE_SEQUENCE
        ]
      };
    }
    module.exports = taggerscript;
  }
});
export default require_taggerscript();
