// @storybook/react/dist/index.mjs
import {
  render,
  renderToDOM
} from "./chunk-7HTWJPRS.mjs";

// src/globals.ts
import global from "global";
var { window: globalWindow } = global;
if (globalWindow) {
  globalWindow.STORYBOOK_ENV = "react";
}

// src/public-api.tsx
import { start } from "@storybook/core-client";
var FRAMEWORK = "react";
var api = start(renderToDOM, { render });
var storiesOf = (kind, m) => {
  return api.clientApi.storiesOf(kind, m).addParameters({
    framework: FRAMEWORK
  });
};
var configure = (...args) => api.configure(FRAMEWORK, ...args);
var addDecorator = api.clientApi.addDecorator;
var addParameters = api.clientApi.addParameters;
var clearDecorators = api.clientApi.clearDecorators;
var setAddon = api.clientApi.setAddon;
var forceReRender = api.forceReRender;
var getStorybook = api.clientApi.getStorybook;
var raw = api.clientApi.raw;

// src/testing-api.ts
import {
  composeStory as originalComposeStory,
  composeStories as originalComposeStories,
  setProjectAnnotations as originalSetProjectAnnotations
} from "@storybook/store";
import { once } from "@storybook/client-logger";
function setProjectAnnotations(projectAnnotations) {
  originalSetProjectAnnotations(projectAnnotations);
}
function setGlobalConfig(projectAnnotations) {
  once.warn(`setGlobalConfig is deprecated. Use setProjectAnnotations instead.`);
  setProjectAnnotations(projectAnnotations);
}
var defaultProjectAnnotations = {
  render
};
function composeStory(story, componentAnnotations, projectAnnotations, exportsName) {
  return originalComposeStory(
    story,
    componentAnnotations,
    projectAnnotations,
    defaultProjectAnnotations,
    exportsName
  );
}
function composeStories(csfExports, projectAnnotations) {
  const composedStories = originalComposeStories(csfExports, projectAnnotations, composeStory);
  return composedStories;
}

// src/index.ts
module?.hot?.decline();
export {
  addDecorator,
  addParameters,
  clearDecorators,
  composeStories,
  composeStory,
  configure,
  forceReRender,
  getStorybook,
  raw,
  setAddon,
  setGlobalConfig,
  setProjectAnnotations,
  storiesOf
};
