// @storybook/components/dist/roboconf-VT6ILN3R.mjs
import {
  __commonJS,
  init_define_process_env
} from "./chunk-NIB2YEFO.mjs";

// ../../node_modules/highlight.js/lib/languages/roboconf.js
var require_roboconf = __commonJS({
  "../../node_modules/highlight.js/lib/languages/roboconf.js"(exports, module) {
    init_define_process_env();
    function roboconf(hljs) {
      const IDENTIFIER = "[a-zA-Z-_][^\\n{]+\\{";
      const PROPERTY = {
        className: "attribute",
        begin: /[a-zA-Z-_]+/,
        end: /\s*:/,
        excludeEnd: true,
        starts: {
          end: ";",
          relevance: 0,
          contains: [
            {
              className: "variable",
              begin: /\.[a-zA-Z-_]+/
            },
            {
              className: "keyword",
              begin: /\(optional\)/
            }
          ]
        }
      };
      return {
        name: "Roboconf",
        aliases: [
          "graph",
          "instances"
        ],
        case_insensitive: true,
        keywords: "import",
        contains: [
          {
            begin: "^facet " + IDENTIFIER,
            end: /\}/,
            keywords: "facet",
            contains: [
              PROPERTY,
              hljs.HASH_COMMENT_MODE
            ]
          },
          {
            begin: "^\\s*instance of " + IDENTIFIER,
            end: /\}/,
            keywords: "name count channels instance-data instance-state instance of",
            illegal: /\S/,
            contains: [
              "self",
              PROPERTY,
              hljs.HASH_COMMENT_MODE
            ]
          },
          {
            begin: "^" + IDENTIFIER,
            end: /\}/,
            contains: [
              PROPERTY,
              hljs.HASH_COMMENT_MODE
            ]
          },
          hljs.HASH_COMMENT_MODE
        ]
      };
    }
    module.exports = roboconf;
  }
});
export default require_roboconf();
