// @storybook/react-webpack5/dist/preset.mjs
var __require = /* @__PURE__ */ ((x) => typeof require !== "undefined" ? require : typeof Proxy !== "undefined" ? new Proxy(x, {
  get: (a, b) => (typeof require !== "undefined" ? require : a)[b]
}) : x)(function(x) {
  if (typeof require !== "undefined")
    return require.apply(this, arguments);
  throw new Error('Dynamic require of "' + x + '" is not supported');
});

// src/preset.ts
import path from "path";
var addons = [
  path.dirname(__require.resolve(path.join("@storybook/preset-react-webpack", "package.json"))),
  path.dirname(__require.resolve(path.join("@storybook/react", "package.json")))
];
var defaultFrameworkOptions = {
  legacyRootApi: true
};
var frameworkOptions = async (_, options) => {
  const config = await options.presets.apply("framework");
  if (typeof config === "string") {
    return {
      name: config,
      options: defaultFrameworkOptions
    };
  }
  if (typeof config === "undefined") {
    return {
      name: __require.resolve("@storybook/react-webpack5"),
      options: defaultFrameworkOptions
    };
  }
  return {
    name: config.name,
    options: {
      ...defaultFrameworkOptions,
      ...config.options
    }
  };
};
var core = async (config, options) => {
  const framework = await options.presets.apply("framework");
  return {
    ...config,
    builder: {
      name: path.dirname(
        __require.resolve(path.join("@storybook/builder-webpack5", "package.json"))
      ),
      options: typeof framework === "string" ? {} : framework.options.builder || {}
    }
  };
};
var webpack = async (config) => {
  config.resolve = config.resolve || {};
  config.resolve.alias = {
    ...config.resolve?.alias,
    "@storybook/react": path.dirname(
      __require.resolve(path.join("@storybook/react", "package.json"))
    )
  };
  return config;
};
export {
  addons,
  core,
  frameworkOptions,
  webpack
};
